lightning/src/ln/chan_utils.rs

```diff
@@ -550,6 +556,415 @@ pub fn build_htlc_transaction(prev_hash: &Txid, feerate_per_kw: u32, contest_del
+			htlc_sigs: Vec::new()
+		}
+	}
+	pub(crate) fn txid<T: secp256k1::Signing + secp256k1::Verification>(&self, channel_static_info: &ChannelStaticInfo, secp_ctx: &Secp256k1<T>) -> Txid {
+		self.info.txid(true, channel_static_info, secp_ctx)
+	}
+}
+
+impl HolderCommitmentTransactionInfo {
+	pub(crate) fn to_holder_commitment_tx<T: secp256k1::Signing + secp256k1::Verification>(&self, channel_static_info: &ChannelStaticInfo, secp_ctx: &Secp256k1<T>) -> HolderCommitmentTransaction {
+		let opt_sigs: Vec<Option<Signature>> = self.htlc_sigs.iter().map(|s| Some(s.clone())).collect();
+		let mut htlcs = self.info.htlcs.clone();
+		let htlcs_with_sig = htlcs.drain(..).zip(opt_sigs).collect();
+		self.info.to_holder_commitment_tx(self.counterparty_sig, htlcs_with_sig, channel_static_info, secp_ctx)
+	}
+}
+
+/// This class tracks the information needed to build a commitment transaction and to actually
+/// build and sign.  It is used for holder transactions that we sign only when needed
+/// and for transactions we sign for the counterparty.
+///
+/// This class can be used inside a signer implementation to generate a signature given the relevant
+/// secret key.
+#[derive(PartialEq)]
+#[derive(Clone)]
+pub struct CommitmentTransactionInfo {
+	/// The backwards-counting commitment number
+	pub commitment_number: u64,
+	/// The value to be sent to the broadcaster
+	pub to_broadcaster_value_sat: u64,
+	/// The value to be sent to the counterparty
+	pub to_countersignatory_value_sat: u64,
+	/// The feerate paid per 1000-weight-unit in this commitment transaction.
+	pub feerate_per_kw: u32,
+	/// The HTLCs which were included in this commitment transaction in output order.
+	pub htlcs: Vec<HTLCOutputInCommitment>,
+	pub(crate) keys: TxCreationKeys,
+}
+
+impl Writeable for Vec<HTLCOutputInCommitment> {
+	#[inline]
+	fn write<W: Writer>(&self, w: &mut W) -> Result<(), ::std::io::Error> {
+		(self.len() as u16).write(w)?;
+		for e in self.iter() {
+			e.write(w)?;
+		}
+		Ok(())
+	}
+}
+
+impl Readable for Vec<HTLCOutputInCommitment> {
+	#[inline]
+	fn read<R: Read>(r: &mut R) -> Result<Self, DecodeError> {
+		let len: u16 = Readable::read(r)?;
+		let byte_size = (len as usize)
+			.checked_mul(HTLC_OUTPUT_IN_COMMITMENT_SIZE)
+			.ok_or(DecodeError::BadLengthDescriptor)?;
+		if byte_size > MAX_BUF_SIZE {
+			return Err(DecodeError::BadLengthDescriptor);
+		}
+		let mut ret = Vec::with_capacity(len as usize);
+		for _ in 0..len { ret.push(HTLCOutputInCommitment::read(r)?); }
+		Ok(ret)
+	}
+}
+
+impl_writeable!(CommitmentTransactionInfo, 0, {
+	commitment_number,
+	to_broadcaster_value_sat,
+	to_countersignatory_value_sat,
+	feerate_per_kw,
+	htlcs,
+	keys
+});
+
+impl CommitmentTransactionInfo {
+	/// Construct an object of the class while assigning transaction output indices to HTLCs.
+	///
+	/// Also keeps track of associated HTLC data and returns it along with the mutated HTLCs.
+	pub fn new_with_auxiliary_htlc_data<T: Copy>(
```

Comment:
the code just accepts the `HTLCSource` items so it can sort them in transaction order as a service to the caller.  it doesn't store it in the constructed object.  note the tuple return value.

I realize this is a bit confusing, but I'm not sure there's a clearer way to go about it.

Reply:
Right, HTLCSource order is a direct dependency of outputs order, so if it's moved outside of this function you will need to sort twice the outputs.